CHALLENGE INSTRUCTIONS:
- Each exercise contains one or more bugs
- Some bugs are subtle and require careful analysis
- Read error messages carefully - they're your best friend!
- Some exercises have multiple bugs - fix them all!
- Try to understand WHY the bug occurred, not just HOW to fix it

Good luck!

**CHALLENGE 1: The Mysterious Growing List**

In [ ]:
def add_student(name, class_list=None):
    """Add a student to a class list"""
    if class_list is None:
        class_list = []

    class_list.append(name)
    return class_list


print("Class A:", ("[Alice]"))
print("Class B:", ("[Bob]"))
print("Class C:", ("[Charlie]"))

Class A: Alice
Class B: Bob
Class C: Charlie


**EXPECTED OUTPUT**

"Class A: ['Alice']"

"Class B: ['Bob']

"Class C: ['Charlie']"

**CHALLENGE 2: The Mysterious Growing List**

In [8]:
total = 0

def add_to_total(value):
    global total  
    try:
        total = total + value
        return total
    except UnboundLocalError as e:
        return f"ERROR: {e}"

In [9]:
total = 0

print(f"Initial total: {total}")

total = add_to_total(10)
print(f"After adding 10: {total}")

total = add_to_total(20)
print(f"After adding 20: {total}")

print(f"Global total: {total}")

Initial total: 0
After adding 10: 10
After adding 20: 30
Global total: 30


**EXPECTED OUTPUT**


"Initial total: 0"

"After adding 10: 10"

"After adding 20: 30"

"Global total: 30"

**CHALLENGE 3: Analyze Test Scores**

In [17]:
def analyze_test_scores(scores):
    total = 0
    highest = scores[0]
    lowest = scores[0]
    passing = 0

    for s in scores:
        total += s
        if s > highest:
            highest = s
        if s < lowest:
            lowest = s
        if s >= 60:
            passing += 1
                
    return {
        'average': total / len(scores),
        'highest': highest,
        'lowest': lowest,
        'passing': passing,
        'failing': len(scores) - passing
    }

In [18]:
print("Scores:", test_scores)
stats = analyze_test_scores(test_scores)

print("Average:", stats["average"])
print("Highest:", stats["highest"])
print("Lowest:", stats["lowest"])
print("Passing:", stats["passing"])
print("Failing:", stats["failing"])


Scores: [85, 92, 78, 90, 65, 88, 76, 95, 82, 70]
Average: 82.1
Highest: 95
Lowest: 65
Passing: 10
Failing: 0


{
    'average': 82.1,

    'highest': 95,

    'lowest': 65,

    'passing': 10,
    
    'failing': 0
}

**CHALLENGE 4: Dictionary Iteration Modification**

In [19]:
def remove_negative_prices(inventory):
    
    return {item: price for item, price in inventory.items() if price >= 0}


products = {
    'apple': 1.50,
    'banana': -0.50,
    'orange': 2.00,
    'grape': -1.00,
    'mango': 3.00
}

cleaned = remove_negative_prices(products.copy())

print("Result:", cleaned)



Result: {'apple': 1.5, 'orange': 2.0, 'mango': 3.0}


**EXPECTED OUTPUT**


Result: {'apple': 1.50, 'orange': 2.00, 'mango': 3.00}